[Yen Ren Zuo Brian] Duke Increments

brianyenna opened the pull request:
Part of Task 4 (Week 3 Individual Project Tasks)

brianyenna committed:
Refactored the code to have initDuke() and terminateDuke() commands.

Purpose of this is separation of concerns: different functions deal with different parts of the Duke code.

brianyenna committed:
Added the runDuke() method. Removed the ASCII art for Duke as it is not part of the Level 1 expected out.

Note that in the future, should the ASCII art be necessary, it can be copied from the original GitHub repo, or obtained from an older commit.

brianyenna committed:
Edited runDuke() method to handle addition of input and listing of input.

All the input is stored in an ArrayList called storedInput (instance variable of Duke)

brianyenna committed:
Implemented Task class

brianyenna committed:
Added ability to mark tasks as 'Done'

brianyenna committed:
Implemented ToDo, Deadline and Event classes.

brianyenna committed:
Added functionality to handle todo, deadline and event

Also updated the text output.

brianyenna committed:
Added comments to code (for documentation)

brianyenna committed:
Added semi-automated testing.

Change to text-ui-test directory and executed ./runtest.sh to begin testing.

brianyenna committed:
Added DukeException class and error-handling code

brianyenna committed:
Updated test input and expected results for text-ui-test

brianyenna committed:
Updated tests for text-ui-test for error-handling (Level 5)

brianyenna committed:
Added support for deletion of Tasks.

Updated tests in text-ui-test to test for deletion functionality as well. The tasks were always stored in an ArrayList, so the current implementation also satisfied A-Collections

brianyenna committed:
Implemented saving functionality

All saving functionality, along with the transition between an ArrayList<Task> and the string storage format in a .txt file is encapsulated within the Storage class.

brianyenna committed:
Implemented DukeDate class to recognise dates in a specific format.

Currently, the date format is in the form dd/mm/yyyy hh:mm. Note that all functionality associated with date parsing and date printing is to be done by the DukeDate class

brianyenna committed:
Made some minor deletions of unnecessary imports

brianyenna committed:
Merge branch 'branch-Level-7'

brianyenna committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

brianyenna committed:
Fixed bug regarding printing of date format.

Bug was due to calling of getSaveFormat() method in parent class Task. When being saved, all tasks (todos, events, deadlines) were being saved as Todos.

brianyenna committed:
Completed Refactoring of code into different classes.

Will need to write docs for the different classes

brianyenna committed:
Re-organised code into packages

brianyenna committed:
Added Units Tests (JUnit)

Tests are for Parser (test all public methods (but not exhaustively)) and for TaskList (only for the size() method)

brianyenna committed:
Made .jar file for A-Jar level.

This commit is not tagged as A-Jar because the release has already been given this title.

brianyenna committed:
Updated javadocs

brianyenna committed:
Updated JavaDocs

brianyenna committed:
Applied Coding Standards

brianyenna committed:
Updated Code Standards

brianyenna committed:
Used IntelliJ to implement coding standards

brianyenna committed:
Implemented Find Functionality

However, note that the find functionality (as of now) is case-sensitive. Not sure whether the task requires a case-insensitive search. In any case, case-insensitive search should not be a lot more work to implement.

brianyenna committed:
Merge branch 'A-JavaDoc'

brianyenna committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/commands/AddCommand.java
#	src/main/java/duke/utils/TaskList.java

brianyenna committed:
Merge branch 'Level-9'

# Conflicts:
#	src/main/java/duke/utils/Parser.java

mrnewguy commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,42 @@
+ * Parent class for all Commands.
+ * Implements the logic that is fundamental to all Commands.
+ */
+public class Command {
```
Well abstracted and good use of inheritance for other command classes.  But is the default behavior really to save the tasks? Is there a better default behavior?

brianyenna replied:
You are definitely right. I was debating over whether this is an ideal default behaviour. The reason I went down this path is because I figured every command ought to result in the file being saved as a safety mechanism. But thinking a little further, this seems wasteful, as some commands never make any modifications, so saving the file is a waste of time and computational resources. 

I'll update this in a future commit. Thanks for pointing it out :)

mrnewguy commented on src/main/java/duke/utils/DukeDate.java:
```diff
@@ -0,0 +1,32 @@
+package duke.utils;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
```
Good use of the SimpleDateFormat. Perhaps you can try writing your own date library too?

mrnewguy commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,128 @@
+ * behaviours. This class is meant as an abstraction to abstract away the details and complexities
+ * behind parsing using input.
+ */
+public class Parser {
```
Good use of switch case statements to keep code concise. However, instead of having multiple parse methods (e.g parseToDo, parseEvent etc), perhaps you could have a parent parse class and have several children inheriting from the parse class. Another method which can shorten this Parser.java class is to have multiple Command Classes (e.g. eventCommand, toDoCommand) instead of just 1 type of AddCommand.

brianyenna replied:
That's definitely a possible approach! I'll look into it. Thank you!

mrnewguy commented on src/test/java/duke/utils/ParserTest.java:
```diff
@@ -0,0 +1,67 @@
+public class ParserTest {
+    @Test
+    public void testParse_Bye_ByeCommand() {
+        try {
```
Don't think there is a need to put try catch statement in your tests. Just running an assert statement with the Junit library should be sufficient

mrnewguy reviewed (commented):
Overall great code, just that I think further inheritance could greatly reduce the length of the Parser.java file.

brianyenna committed:
First commit after setting up of Gradle

brianyenna committed:
Updated Gradle Config for Build Automation

brianyenna committed:
Merge branch 'A-Gradle'

brianyenna committed:
Implement Checkstyle via Gradle

brianyenna committed:
Apply changes to adhere to coding style

brianyenna committed:
Further changes to adhere to coding style.

brianyenna committed:
Change to adhere to coding style.

This is in accordance to the generated .html file after running gradle checkstyleMain and gradle checkstyleTest

brianyenna committed:
Made config/checkstyle/checkstyle.xml file

brianyenna committed:
Merge branch 'A-CheckStyle'

brianyenna committed:
Add JavaFX functionality.

Edit build.gradle to add JavaFX functionality.

Duke's constructor had to be changed to take in no arguments. Previously, a String representing the saveFilePath was passed into the constructor, but this had to be changed to accommodate the JavaFX launch code. Now, the saveFilePath can still be edited, but any change has to be made directly to the static variable, instead of being passed to the Duke constructor directly. An example is given in the main() method of Duke.

gachia commented:
Excellent usage of inheritance and abstraction. Quite well done.

gachia commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,26 @@
+
+        //Parent method invoked to save TaskList
+        super.execute(ui, storage, allTasks);
+    }
```
Curious to know if the ByeCommand is able to execute and exit the program if a file path is not specified or broken. I encountered it on my code and yours look similar.

brianyenna replied:
Hmm I didn't encounter this issue in my code. For me, the file path is saved as a static variable, so it does not change once the application starts. Could you clarify on the issue you are facing? Do you mean to say that pass your ByeCommand object a file path?

gachia replied:
It's more of an edge case test, where the file path does not exist or is broken when the application starts. So when you try to save, a DukeException gets thrown and you aren't able to exit the program with a ByeCommand because the boolean variable was unable to set to True.

brianyenna committed:
Complete Part 2 of JavaFX Tutorial

brianyenna committed:
Complete Part 3 of JavaFX Tutorial

brianyenna committed:
Complete Part 4 of JavaFX Tutorial

brianyenna committed:
Merge branch 'A-JavaFx'

brianyenna committed:
Add CSS styling to elements.

brianyenna committed:
Merge tag 'Level-10'

brianyenna committed:
Make coding style changes to adhere to coding style standards.

brianyenna committed:
Add .travis.yml file

brianyenna committed:
Update .travis.yml file

brianyenna committed:
Merge branch 'A-Travis'

brianyenna committed:
Update with assert statements. Change build.gradle file to enable assertions when `run` is invoked by gradle.

brianyenna committed:
Add assert statements

brianyenna committed:
Add assert statement for DukeDate object creation

brianyenna committed:
Refactor Command and its subclasses.

Original Command#execute would save the TaskList object. Refactored to remove this functionality as not all subclasses require this behaviour. Makes more sense to explicitly move the saving functionality to the subclasses instead.

brianyenna committed:
Updated Parser. Replace ByeCommand with SaveCommand as "bye" no longer closes the window. Instead, typing "save" will not explicitly save the data.

Note that this is not technically a refactoring as external behaviour is changed.

brianyenna committed:
Update tests to reflect change from ByeCommand to SaveCommand.

brianyenna committed:
Refactor Command class to get rid of dead code.

More specifically, the isExit boolean is no longer necessary. It was initially meant as a flag to indicate whether the CLI Duke ought to exit from the while loop, but in the GUI Duke, this functionality is no longer supported.

brianyenna committed:
Refactor Dialogbox.

Original Dialogbox had a boolean to determine whether to flip the dialogbox (user vs duke) and which fxml file to load. Refactoring Dialogbox into 2 different classes - UserDialogBox and DukeDialogBox - is cleaner, since there already exist 2 different fxml files detailing their layouts and their respectively css style classes. This is in line with the 'Remove Flag Argument' refactoring principle stated on refactoring.com/catalog/

brianyenna committed:
Restructured GUI code.

DukeDialogbox, UserDialogbox and MainWindow are now in package duke.gui.

brianyenna committed:
Fix assert statement bug.

brianyenna committed:
Merge pull request #4 from brianyenna/A-Assertions

A-Assertions

brianyenna committed:
Merge branch 'master' of github.com:brianyenna/duke

brianyenna committed:
Merge branch 'master' into A-CodeQuality

# Conflicts:
#	data/savedTasks.txt

brianyenna committed:
Merge pull request #5 from brianyenna/A-CodeQuality

A-CodeQuality

brianyenna committed:
Merge branch 'master' of github.com:brianyenna/duke

brianyenna committed:
Add help command and fix issue about truncation of text when the text is too tall in the GUI.

brianyenna committed:
Change javafx Label to Text

brianyenna committed:
Merge branch 'C-Help'

brianyenna committed:
Update docs/README.md to include user guide.

brianyenna committed:
Update wrapping and text alignment.

brianyenna committed:
Add Ui.png and some basic styling (font)

brianyenna committed:
Merge branch 'A-UserGuide'

brianyenna committed:
Update dependencies in build.gradle

brianyenna committed:
Update build.gradle

brianyenna committed:
Update Title

brianyenna committed:
Update Ui.png

brianyenna committed:
Set theme jekyll-theme-minimal

brianyenna committed:
Fix Storage#save bug that prevented command recognition in a JAR file.

brianyenna committed:
Merge branch 'fix-jar-file'

brianyenna committed:
Merge branches 'master' and 'master' of github.com:brianyenna/duke

brianyenna committed:
Update README.md

Update typos and minor inaccuracies in README.md

brianyenna committed:
Update README.md

Update README.md with addition of 'done' command details

brianyenna committed:
Merge branches 'master' and 'master' of github.com:brianyenna/duke

brianyenna committed:
Update README.md (typos)
